packages/vite/src/node/server/middlewares/indexHtml.ts

```diff
@@ -82,6 +86,15 @@ function getHtmlFilename(url: string, server: ViteDevServer) {
   }
 }
 
+function shouldPreTransform(url: string, config: ResolvedConfig) {
+  return (
+    !checkPublicFile(url, config) &&
+    !isExternalUrl(url) &&
+    !isDataUrl(url) &&
```

Comment:
I think the code paths that would call `preTransformRequest` wouldn't hit an external url or data url, so we can skip this. Primarily the `processNodeUrl` should have the conditions that avoid them.

Scoping down the JS and CSS requests sounds good to me though. I'm hoping these additions still make the optimization worth while, otherwise it might be better off letting the browser request them to keep it simple. 🤔

Reply:
Done at [`08f4e6f` (#13231)](https://github.com/vitejs/vite/pull/13231/commits/08f4e6fecc9f4e36a1d2a645905c34f4d8d71af2) 👍🏼